src/test/scala-3/com/github/tminglei/slickpg/PgCompositeSupportSuite.scala

```diff
@@ -0,0 +1,317 @@
+package com.github.tminglei.slickpg
+
+import java.time.LocalDateTime
+import scala.concurrent.Await
+import scala.concurrent.duration._
+import scala.collection.immutable.{Range => ScalaRange}
+import scala.jdk.CollectionConverters._
+import slick.jdbc.{GetResult, PositionedResult, PostgresProfile, SetParameter}
+import com.github.tminglei.slickpg.composite.Struct
+import org.postgresql.util.HStoreConverter
+import org.scalatest.funsuite.AnyFunSuite
+
+import com.github.tminglei.slickpg.utils.{RegisteredTypeConverter,TypeConverters}
+
+
+object PgCompositeSupportSuite {
+
+  import TypeConverters.*
+  def ts(str: String) = LocalDateTime.parse(str.replace(' ', 'T'))
+
+  case class Composite1(
+    id: Long,
+    txt: String,
+    date: LocalDateTime,
+    tsRange: Option[Range[LocalDateTime]]
+  ) extends Struct
+
+  case class Composite2(
+    id: Long,
+    comp1: Composite1,
+    confirm: Boolean,
+    map: Map[String, String]
+  ) extends Struct
+
+  case class Composite3(
+    name: Option[String] = None,
+    code: Option[Int] = None,
+    num: Option[Int] = None,
+    bool: Option[Boolean] = None
+  ) extends Struct
+
+  case class Composite4(
+    id: Int,
+    name: String,
+    variable: List[String],
+    optVars: Option[List[String]],
+    script1: String,
+    script2: String
+  ) extends Struct
+
+  case class C1 (
+    name: String
+  ) extends Struct
+
+  case class C2 (
+    c1: List[C1]
+  ) extends Struct
+
+  //-------------------------------------------------------------
+  trait MyPostgresProfile1 extends PostgresProfile with PgCompositeSupport with PgArraySupport with utils.PgCommonJdbcTypes {
+    def mapToString(m: Map[String, String]): String = HStoreConverter.toString((m).asJava)
+    def stringToMap(s: String): Map[String, String] = (HStoreConverter.fromString(s)
+      .asInstanceOf[java.util.Map[String, String]]).asScala.toMap
+
+    ///
+    trait API extends JdbcAPI with ArrayImplicits {
+      implicit val StringToRange: RegisteredTypeConverter[String, Range[LocalDateTime]] =
+        RegisteredTypeConverter(PgRangeSupportUtils.mkRangeFn(ts))
+      implicit val RangeToString: RegisteredTypeConverter[Range[LocalDateTime], String] =
+        RegisteredTypeConverter(PgRangeSupportUtils.toStringFn[LocalDateTime](_.toString))
+      implicit val MapToString: RegisteredTypeConverter[Map[String, String], String] = RegisteredTypeConverter(mapToString)
+      implicit val StringToMap: RegisteredTypeConverter[String, Map[String, String]] = RegisteredTypeConverter(stringToMap)
```

Comment:
So this is different to scala 2 usage -- rather than 'register'ing a typetag, we need to have an appropriate implicit 'registered converter' in scope. This all boils down to the composite type being constructed at compile time rather than runtime, and I _think_ it's actually an improvement. Sadly it does mean some inconsistency between scala 2 and scala 3 APIs, because we can't use typetags in scala 3 and can't use inline in scala 2, but hey ho them's the breaks